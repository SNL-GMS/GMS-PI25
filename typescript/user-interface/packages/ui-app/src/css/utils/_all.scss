@import './variables.scss';
@import './string-utils.scss';
@import './layout-utils.scss';
@import './number-utils.scss';
@import './visual-utils.scss';
@import './interaction-utils.scss';
@import './table.scss';
