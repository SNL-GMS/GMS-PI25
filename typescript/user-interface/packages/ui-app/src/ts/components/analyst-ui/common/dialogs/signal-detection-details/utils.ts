import { FilterTypes, SignalDetectionTypes } from '@gms/common-model';
import {
  DATE_TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION,
  dateToString,
  toDate
} from '@gms/common-util';
import { FormTypes } from '@gms/ui-core-components';
import flatten from 'lodash/flatten';

import { parseWaveformChannelType } from '~analyst-ui/common/utils/signal-detection-util';
import { messageConfig } from '~analyst-ui/config/message-config';
import { formatNumberForDisplayFixedThreeDecimalPlaces } from '~common-ui/common/table-utils';

import type { SignalDetectionHistoryRow } from './types';

/**
 * Generate the table row data for the detection history.
 */
export function generateDetectionHistoryTableRows(
  detection: SignalDetectionTypes.SignalDetection
): SignalDetectionHistoryRow[] {
  const rows = flatten(
    detection.signalDetectionHypotheses
      .map(signalDetectionHypo => {
        const phaseFMValue = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
          signalDetectionHypo?.featureMeasurements
        );
        const arrivalTimeFM = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(
          signalDetectionHypo?.featureMeasurements
        );
        const arrivalTimeFMValue = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
          signalDetectionHypo?.featureMeasurements
        );
        return {
          id: signalDetectionHypo.id.id,
          versionId: signalDetectionHypo.id.id, // TODO: what should the version be? (hypo number?)
          creationTimestamp: -1, // TODO: change when field is available for now -1 translates to 'TBD'
          phase: phaseFMValue.value,
          arrivalTimeMeasurementFeatureType:
            SignalDetectionTypes.FeatureMeasurementType.ARRIVAL_TIME,
          arrivalTimeMeasurementTimestamp: arrivalTimeFMValue.arrivalTime.value,
          arrivalTimeMeasurementUncertaintySec: arrivalTimeFMValue.arrivalTime.standardDeviation,
          channelSegmentType: parseWaveformChannelType(arrivalTimeFM.channel.name),
          authorName: 'TBD',
          deleted: signalDetectionHypo?.deleted ? 'True' : 'False'
        };
      })
      .sort((a, b) => b.arrivalTimeMeasurementTimestamp - a.arrivalTimeMeasurementTimestamp)
  );
  rows[0]['first-in-table'] = true;
  return rows;
}

/**
 * Given a signalDetection, build a list of {@link FormTypes.FormItem}s to be used
 * in the Signal Detection Details menu
 *
 * @param signalDetection Signal Detection to build against
 * @param filterDefinitionByFilterDefinitionUsage Optional filter definitions being used
 * @returns
 */
export function buildSdDetailFormItems(
  signalDetection: SignalDetectionTypes.SignalDetection,
  filterDefinitionByFilterDefinitionUsage?: FilterTypes.FilterDefinitionByFilterDefinitionUsage
): FormTypes.FormItem[] {
  const formItems: FormTypes.FormItem[] = [];
  const sdHypothesis = SignalDetectionTypes.Util.getCurrentHypothesis(
    signalDetection.signalDetectionHypotheses
  );
  const arrivalTimeFM = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(
    sdHypothesis?.featureMeasurements
  );
  const arrivalTimeFMValue = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
    sdHypothesis?.featureMeasurements
  );
  const fmPhase = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
    sdHypothesis?.featureMeasurements
  );

  // Build the arrival uncertainty formatted string
  let arrivalTimeUncertainty = formatNumberForDisplayFixedThreeDecimalPlaces(
    arrivalTimeFMValue.arrivalTime.standardDeviation
  );
  // If a valid fields add 's' seconds unit
  if (arrivalTimeUncertainty !== messageConfig.invalidCellText) {
    arrivalTimeUncertainty = `${arrivalTimeUncertainty}s`;
  }

  formItems.push(
    {
      itemKey: 'Station',
      labelText: 'Station',
      itemType: FormTypes.ItemType.Display,
      displayText: signalDetection.station.name
    },
    {
      itemKey: 'Creation time',
      labelText: 'Creation time',
      itemType: FormTypes.ItemType.Display,
      displayText: 'TBD'
      // displayTextFormat: FormTypes.TextFormats.Time
    },
    {
      itemKey: 'Phase',
      labelText: 'Phase',
      displayText: fmPhase.value ?? 'Unknown',
      itemType: FormTypes.ItemType.Display
    },
    {
      itemKey: 'Detection time',
      labelText: 'Detection time',
      itemType: FormTypes.ItemType.Display,
      className: 'monospace',
      displayText: `${dateToString(
        toDate(arrivalTimeFMValue?.arrivalTime.value),
        DATE_TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION
      )} Â± ${arrivalTimeUncertainty}`,
      displayTextFormat: FormTypes.TextFormats.Time
    },
    {
      itemKey: 'Channel segment type',
      labelText: 'Channel segment type',
      itemType: FormTypes.ItemType.Display,
      displayText: parseWaveformChannelType(arrivalTimeFM.channel.name)
    },
    {
      itemKey: 'Author',
      labelText: 'Author',
      itemType: FormTypes.ItemType.Display,
      displayText: 'TBD'
    },
    {
      itemKey: 'Deleted',
      labelText: 'Deleted',
      itemType: FormTypes.ItemType.Display,
      displayText: SignalDetectionTypes.Util.getCurrentHypothesis(
        signalDetection.signalDetectionHypotheses
      )?.deleted
        ? 'True'
        : 'False'
    }
  );

  if (filterDefinitionByFilterDefinitionUsage) {
    formItems.push(
      {
        itemKey: 'Detection Filter',
        labelText: 'Detection Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.DETECTION].name
      },
      {
        itemKey: 'Onset Filter',
        labelText: 'Onset Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.ONSET].name
      },
      {
        itemKey: 'FK Filter',
        labelText: 'FK Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.FK].name
      }
    );
  }

  return formItems;
}
