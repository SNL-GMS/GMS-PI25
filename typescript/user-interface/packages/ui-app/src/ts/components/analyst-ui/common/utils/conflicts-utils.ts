import { EventTypes, SignalDetectionTypes } from '@gms/common-model';
import type { AssociationConflictRecord } from '@gms/ui-state';

import type { ConflictedEventInfo, ConflictedSDInfo } from '../tooltips/conflicts-tooltip-content';

/**
 * Given two objects with a `time` property, sort them in chronological order.
 */
export function sortByTimeChronological(a: { time: number }, b: { time: number }) {
  if (a.time < b.time) return -1;
  if (a.time > b.time) return 1;
  return 0;
}

/**
 *
 * @param allSignalDetectionsInInterval
 * @param allEventsInInterval
 * @param allConflicts
 * @param openIntervalName
 * @param currentEventId
 * @returns the array of signal detections with conflicts associated to the input event
 */
export const getConflictedSDsForEvent = (
  allSignalDetectionsInInterval: SignalDetectionTypes.SignalDetection[],
  allEventsInInterval: EventTypes.Event[],
  allConflicts: AssociationConflictRecord,
  openIntervalName: string,
  currentEventId: string
): ConflictedSDInfo[] => {
  // current event of the row
  const currentEvent: EventTypes.Event = allEventsInInterval?.find(ev => ev.id === currentEventId);
  if (!currentEvent) return [];

  let currentEventHypothesis: EventTypes.EventHypothesis = EventTypes.findPreferredEventHypothesisByStage(
    currentEvent,
    openIntervalName
  );
  if (
    currentEventHypothesis === undefined ||
    currentEventHypothesis.locationSolutions === undefined
  ) {
    currentEventHypothesis = EventTypes.findEventHypothesisParent(
      currentEvent,
      currentEventHypothesis
    );
  }

  const currentEventAssociatedSDIds: string[] =
    currentEventHypothesis?.associatedSignalDetectionHypotheses.map(
      sd => sd.id.signalDetectionId
    ) || [];

  // filter the array to only have sd id's that are associated to the current event
  const associatedSDIdsWithConflicts: string[] =
    currentEventAssociatedSDIds.length > 0
      ? Object.keys(allConflicts).filter(sdid => currentEventAssociatedSDIds.includes(sdid))
      : [];

  const sdConflicts =
    associatedSDIdsWithConflicts.length > 0
      ? allSignalDetectionsInInterval
          ?.filter(sd => associatedSDIdsWithConflicts.includes(sd.id)) // only the sd's associated to the current event with conflicts
          .map<ConflictedSDInfo>(sd => {
            const sdHypothesis = SignalDetectionTypes.Util.getCurrentHypothesis(
              sd.signalDetectionHypotheses
            );
            const sdTime: number = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
              sdHypothesis.featureMeasurements
            )?.arrivalTime?.value;
            const sdPhase = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
              sdHypothesis.featureMeasurements
            )?.value;

            return { station: sd.station.name, phase: sdPhase, time: sdTime };
          })
      : [];

  if (sdConflicts?.length > 1) {
    sdConflicts.sort(sortByTimeChronological);
  }
  return sdConflicts;
};

/**
 * @param allEventsInInterval
 * @param allConflicts
 * @param openIntervalName
 * @param currentSDId
 * @returns the array of events with conflicts because the input SD is signal detections with conflicts associated to the input event
 */
export const getConflictedEventsForSD = (
  allEventsInInterval: EventTypes.Event[],
  allConflicts: AssociationConflictRecord,
  openIntervalName: string,
  currentSDId: string
): ConflictedEventInfo[] => {
  const eventConflicts =
    allConflicts[currentSDId]?.eventHypothesisIds
      .map<ConflictedEventInfo>(ev => {
        const currentEvent = allEventsInInterval?.find(evt => evt.id === ev.eventId);
        if (!currentEvent) return undefined;

        let currentEventHypothesis: EventTypes.EventHypothesis = EventTypes.findPreferredEventHypothesisByStage(
          currentEvent,
          openIntervalName
        );

        // If no preferredHypothesis then default to the parent hypothesis
        if (
          currentEventHypothesis === undefined ||
          currentEventHypothesis.locationSolutions === undefined
        ) {
          currentEventHypothesis = EventTypes.findEventHypothesisParent(
            currentEvent,
            currentEventHypothesis
          );
        }

        if (currentEventHypothesis) {
          const locationSolution = EventTypes.findPreferredLocationSolution(
            currentEventHypothesis.id.hypothesisId,
            currentEvent.eventHypotheses
          );
          return { time: locationSolution?.location?.time };
        }
        return undefined;
      })
      .filter(e => e !== undefined) || [];

  if (eventConflicts?.length > 1) {
    eventConflicts.sort(sortByTimeChronological);
  }
  return eventConflicts;
};
