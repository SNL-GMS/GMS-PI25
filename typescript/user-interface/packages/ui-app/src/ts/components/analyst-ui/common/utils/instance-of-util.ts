/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import type { FilterTypes, FkTypes } from '@gms/common-model';
import { ChannelSegmentTypes } from '@gms/common-model';
import { UNFILTERED_FILTER } from '@gms/common-model/lib/filter/types';

/**
 * Checks if FK spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
export function isFkSpectraChannelSegment(
  object: ChannelSegmentTypes.ChannelSegment<ChannelSegmentTypes.TimeSeries>
): object is ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra> {
  return object.timeseriesType === ChannelSegmentTypes.TimeSeriesType.FK_SPECTRA;
}

/**
 * Creates/Returns an unfiltered waveform filter
 */
export function createUnfilteredWaveformFilter(): FilterTypes.Filter {
  return UNFILTERED_FILTER;
}
