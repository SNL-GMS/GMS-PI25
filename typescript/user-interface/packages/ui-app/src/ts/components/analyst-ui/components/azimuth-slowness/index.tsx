import { IanDisplays, toDisplayTitle } from '@gms/common-model/lib/displays/types';
import { nonIdealStateWithSpinner } from '@gms/ui-core-components';
import React from 'react';

import type { AzimuthSlownessProps } from './types';

const AzimuthSlownessLazy = React.lazy(async () =>
  import(
    /* webpackChunkName: 'ui-app-azimuth-slowness' */ './azimuth-slowness-container'
  ).then(module => ({ default: module.AzimuthSlowness }))
);

export function AzimuthSlowness(props: AzimuthSlownessProps) {
  return (
    <React.Suspense
      fallback={nonIdealStateWithSpinner(
        toDisplayTitle(IanDisplays.AZIMUTH_SLOWNESS),
        'Loading Display'
      )}
    >
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <AzimuthSlownessLazy {...props} />
    </React.Suspense>
  );
}
