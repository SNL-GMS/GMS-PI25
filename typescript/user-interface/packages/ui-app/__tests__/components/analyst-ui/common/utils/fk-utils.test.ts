import type { FkTypes, StationTypes } from '@gms/common-model';
import { SignalDetectionTypes } from '@gms/common-model';
import { eventData, signalDetectionsData } from '@gms/common-model/__tests__/__data__';
import { getTestFkChannelSegmentRecord, getTestFkData } from '@gms/ui-state/__tests__/__data__';

import { createChannelSegmentString } from '~analyst-ui/common/utils/signal-detection-util';

import {
  calculateStartTimeForFk,
  computeFk,
  createComputeFkInput,
  frequencyBandToString,
  getAssociatedDetectionsWithFks,
  getDefaultFkConfigurationForSignalDetection,
  getFkChannelSegment,
  getFkData,
  getFkParams
} from '../../../../../src/ts/components/analyst-ui/common/utils/fk-utils';
import { fkInput, newConfiguration } from '../../../../__data__/azimuth-slowness';

// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

const arrivalTimeFm = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(
  signalDetectionsData[0].signalDetectionHypotheses[0].featureMeasurements
);
const azFm = SignalDetectionTypes.Util.findAzimuthFeatureMeasurement(
  signalDetectionsData[0].signalDetectionHypotheses[0].featureMeasurements
);
const fk = getTestFkData(arrivalTimeFm.measurementValue.arrivalTime.value);
const fkChannelSegmentRecord = getTestFkChannelSegmentRecord(signalDetectionsData[0]);

/**
 * Tests the ability to check if the peak trough is in warning
 */
describe('frequencyBandToString', () => {
  test('correctly creates frequency band string', () => {
    const band: FkTypes.FrequencyBand = {
      maxFrequencyHz: 5,
      minFrequencyHz: 1
    };
    const testString = '1 - 5 Hz';
    expect(frequencyBandToString(band)).toEqual(testString);
  });
});

describe('Can retrieve FkData', () => {
  test('cs string', () => {
    const csString = createChannelSegmentString(azFm.measuredChannelSegment.id);
    expect(csString).toBeDefined();
  });
  test('getFkData', () => {
    expect(arrivalTimeFm).toBeDefined();
    expect(getFkData(undefined, fkChannelSegmentRecord)).toBeUndefined();
    expect(getFkData(signalDetectionsData[0], fkChannelSegmentRecord)).toBeDefined();
  });

  test('getFkChannelSegment', () => {
    expect(arrivalTimeFm).toBeDefined();
    expect(getFkChannelSegment(undefined, fkChannelSegmentRecord)).toBeUndefined();
    expect(getFkChannelSegment(signalDetectionsData[0], fkChannelSegmentRecord)).toBeDefined();
  });

  test('createComputeFkInput', () => {
    const fkParams = getFkParams(fk);
    expect(
      createComputeFkInput(undefined, fkParams, newConfiguration, false, fkChannelSegmentRecord)
    ).toBeUndefined();
    expect(
      createComputeFkInput(
        signalDetectionsData[0],
        fkParams,
        newConfiguration,
        false,
        fkChannelSegmentRecord
      )
    ).toMatchSnapshot();
  });
  test('computeFk', () => {
    const dispatch = jest.fn();
    expect(computeFk(fkInput, dispatch)).toBeUndefined();
    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  test('getFkParamsForSd', () => {
    expect(getFkParams(undefined)).toBeUndefined();
    expect(getFkParams(fk)).toBeDefined();
  });

  describe('General Fk configuration and SD associations', () => {
    test('getAssociatedDetectionsWithFks', () => {
      const openIntervalName = 'AL1';
      const eventHypos = eventData.eventHypotheses;
      const sdHypo = SignalDetectionTypes.Util.getCurrentHypothesis(
        signalDetectionsData[0].signalDetectionHypotheses
      );
      eventHypos[eventHypos.length - 1].associatedSignalDetectionHypotheses.push(sdHypo);
      expect(
        getAssociatedDetectionsWithFks(
          undefined,
          signalDetectionsData,
          fkChannelSegmentRecord,
          openIntervalName
        )
      ).toHaveLength(0);
      expect(
        getAssociatedDetectionsWithFks(
          eventData,
          undefined,
          fkChannelSegmentRecord,
          openIntervalName
        )
      ).toHaveLength(0);
      expect(
        getAssociatedDetectionsWithFks(eventData, [], fkChannelSegmentRecord, openIntervalName)
      ).toHaveLength(0);
      expect(
        getAssociatedDetectionsWithFks(
          eventData,
          signalDetectionsData,
          fkChannelSegmentRecord,
          openIntervalName
        )
      ).toHaveLength(1);
    });
  });

  test('getDefaultFkConfigurationForSignalDetection', () => {
    const station = {
      name: 'ASAR',
      effectiveAt: null
    };
    expect(
      getDefaultFkConfigurationForSignalDetection(
        signalDetectionsData[0],
        station as StationTypes.Station,
        getFkData(signalDetectionsData[0], fkChannelSegmentRecord)
      )
    ).toMatchSnapshot();
  });

  test('calculateStartTimeForFk', () => {
    let startTime = 130;
    const arrivalTime = 120;
    const leadTime = 1;
    const stepSize = 2;
    expect(calculateStartTimeForFk(startTime, arrivalTime, leadTime, undefined)).toBeUndefined();
    expect(calculateStartTimeForFk(startTime, arrivalTime, leadTime, stepSize)).toBeUndefined();
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    startTime = 100;
    expect(
      calculateStartTimeForFk(startTime, arrivalTime, leadTime, stepSize)
    ).toMatchInlineSnapshot(`101`);
  });
});
