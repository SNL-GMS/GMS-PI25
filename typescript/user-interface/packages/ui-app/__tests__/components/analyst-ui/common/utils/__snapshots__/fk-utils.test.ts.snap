// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can retrieve FkData createComputeFkInput 1`] = `
{
  "configuration": {
    "contributingChannelsConfiguration": [
      {
        "enabled": true,
        "id": "ARCES.ARA0.BHE",
        "name": "ARCES.ARA0.BHE",
      },
      {
        "enabled": true,
        "id": "ARCES.ARA0.BHN",
        "name": "ARCES.ARA0.BHN",
      },
      {
        "enabled": true,
        "id": "ARCES.ARA0.BHZ",
        "name": "ARCES.ARA0.BHZ",
      },
      {
        "enabled": true,
        "id": "ARCES.ARA1.BHE",
        "name": "ARCES.ARA1.BHE",
      },
      {
        "enabled": true,
        "id": "ARCES.ARA1.BHN",
        "name": "ARCES.ARA1.BHN",
      },
    ],
    "leadFkSpectrumSeconds": 1,
    "maximumSlowness": 40,
    "mediumVelocity": 5.8,
    "normalizeWaveforms": false,
    "numberOfPoints": 81,
    "useChannelVerticalOffset": false,
  },
  "fkComputeInput": {
    "channels": [
      {
        "name": "ARCES.ARA0.BHE",
      },
      {
        "name": "ARCES.ARA0.BHN",
      },
      {
        "name": "ARCES.ARA0.BHZ",
      },
      {
        "name": "ARCES.ARA1.BHE",
      },
      {
        "name": "ARCES.ARA1.BHN",
      },
    ],
    "highFrequency": 2,
    "lowFrequency": 0.5,
    "normalizeWaveforms": false,
    "phaseType": "P",
    "sampleCount": 604,
    "sampleRate": 2,
    "slowCountX": 81,
    "slowCountY": 81,
    "slowDeltaX": 0.008882188700431906,
    "slowDeltaY": 0.008882188700431906,
    "slowStartX": -0.3597286423674922,
    "slowStartY": -0.3597286423674922,
    "startTime": undefined,
    "useChannelVerticalOffset": false,
    "windowLead": "PT1S",
    "windowLength": "PT5S",
  },
  "isThumbnailRequest": false,
  "signalDetectionId": "012de1b9-8ae3-3fd4-800d-58665c3152cc",
  "windowParams": {
    "leadSeconds": 1,
    "lengthSeconds": 5,
    "stepSize": 0.5,
  },
}
`;

exports[`Can retrieve FkData getDefaultFkConfigurationForSignalDetection 1`] = `
{
  "fkSpectrumDurationSeconds": 300,
  "frequencyBand": {
    "maxFrequencyHz": 2,
    "minFrequencyHz": 0.5,
  },
  "leadFkSpectrumSeconds": 1,
  "maximumSlowness": 40,
  "mediumVelocity": 5.8,
  "normalizeWaveforms": false,
  "numberOfPoints": 81,
  "prefilter": {
    "filterDefinition": null,
    "namedFilter": null,
    "unfiltered": true,
    "withinHotKeyCycle": true,
  },
  "selectedChannels": [],
  "stepSizeSeconds": 0.5,
  "useChannelVerticalOffset": false,
  "window": {
    "durationSecs": 5,
    "leadSecs": 1,
  },
}
`;
